config = {
    # need to use refseq since we need to match the transcript IDs in the processed reactivity data
    'genome_annotation': 'ucsc_refseq.2017-06-25',


    # # cell line gene expression
    # # can be used to subset to highly expressed genes
    # 'cell_line_gene_expression': 'Pnb4DN',
    # 'cell_line': 'HEK 293',
    # 'min_tpm': 20,

    # 'reactivity_min': 0.0,
    # 'reactivity_max': 1.0,
    # 'reactivity_clip': False,

    # dataset
    # generated by https://github.com/PSI-Lab/alice-sandbox/tree/258b57264f8987fd146f8e7936c7a4e8892cd32b/rna_ss/data_processing/combine_data
    'gtrack': 'data/reactivity_combined.gtrack',
    'all_inervals': 'data/intervals_combined.txt',

    # # all chromosomes used for training and validation
    # # will be split into 5 folds
    # 'chrom_all': ['chr%d' % x for x in range(1, 23)],
    # # chrom sizes, from
    # 'chrom_sizes': [['chr1', 249250621],
    #                 ['chr2', 243199373],
    #                 ['chr3', 198022430],
    #                 ['chr4', 191154276],
    #                 ['chr5', 180915260],
    #                 ['chr6', 171115067],
    #                 ['chr7', 159138663],
    #                 ['chr8', 146364022],
    #                 ['chr9', 141213431],
    #                 ['chr10', 135534747],
    #                 ['chr11', 135006516],
    #                 ['chr12', 133851895],
    #                 ['chr13', 115169878],
    #                 ['chr14', 107349540],
    #                 ['chr15', 102531392],
    #                 ['chr16', 90354753],
    #                 ['chr17', 81195210],
    #                 ['chr18', 78077248],
    #                 ['chr20', 63025520],
    #                 ['chr19', 59128983],
    #                 ['chr22', 51304566],
    #                 ['chr21', 48129895]],

    # chromosome for different folds
    'chrom_folds': [['chr1', 'chr10', 'chr15', 'chr20', 'chr22'],
                    ['chr2', 'chr9', 'chr14', 'chr19', 'chr21'],
                    ['chr3', 'chr8', 'chr11', 'chr17'],
                    ['chr4', 'chr7', 'chr12', 'chr18'],
                    ['chr5', 'chr6', 'chr13', 'chr16']],

    # # save disjoint intervals for each of the 5 folds
    # 'interval_folds': [
    #     'data/intervals_fold_0.txt',
    #     'data/intervals_fold_1.txt',
    #     'data/intervals_fold_2.txt',
    #     'data/intervals_fold_3.txt',
    #     'data/intervals_fold_4.txt',
    # ],

    # model
    'n_filters': 32,  # n filter in first layer
    'residual_conv': [
        {'num_filter': 32, 'filter_width': 11, 'dilation': 1},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 1},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 1},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 1},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 4},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 4},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 4},
        {'num_filter': 32, 'filter_width': 11, 'dilation': 4},
        # {'num_filter': 32, 'filter_width': 21, 'dilation': 10},
        # {'num_filter': 32, 'filter_width': 21, 'dilation': 10},
        # {'num_filter': 32, 'filter_width': 21, 'dilation': 10},
        # {'num_filter': 32, 'filter_width': 21, 'dilation': 10},
        # {'num_filter': 32, 'filter_width': 41, 'dilation': 25},
        # {'num_filter': 32, 'filter_width': 41, 'dilation': 25},
        # {'num_filter': 32, 'filter_width': 41, 'dilation': 25},
        # {'num_filter': 32, 'filter_width': 41, 'dilation': 25},
    ],


    'n_repeat_in_residual_unit': 2,
    'skip_conn_every_n': 4,

    'example_length': 500,
    'batch_size': 10,
    'learning_rate': 0.00001,
    'residual': True,
    'skipconn': True,
    'gated': False,

    'num_epoch': 200,
    'num_batch_for_validation': 10,
    'es_patience': 10,

    # final models
    'model_dir': 'model/',

}
